### ISSUE TITLE
Apply source presence tracking recursively

### ISSUE THREAD

**Issue by birdfriend** [mapstruct/mapstruct#1456]
Hi,

We want to apply source presence tracking recursively. For example:
//source object
```
public class SourceA {
  private String fieldA;
  private SourceB fieldB;

//getters & setters & hasXXX methods
...
}

public class SourceB {
  private String fieldC;
//getters & setters & hasXXX methods
...
}

//target
public class TargetObj {
  private String fieldA;
  private String fieldC;
//getters & setters
...
}
```
//mapper
```
@Mapper(nullValueCheckStrategy =  NullValueCheckStrategy.ALWAYS)
public interface TestMapper {

  @Mapping(source = "fieldB.fieldC", target = "fieldC")
  TargetObj testMap(SourceA source);
}
```

//The auto generated code:
```
public TargetObj testMap(SourceA source) {
    if(source == null) {
      return null;
    } else {
      TargetObj targetObj = new TargetObj();
      String fieldC = this.sourceFieldBFieldC(source);
      if(source.hasFieldB()) {
        targetObj.setFieldC(fieldC);
      }

      if(source.hasFieldA()) {
        targetObj.setFieldA(source.getFieldA());
      }

      return targetObj;
    }
  }
...
```

When mapping `fieldC`, source presence tracking has only applied to it's parent object,
```
if(source.hasFieldB()) {
        targetObj.setFieldC(fieldC);
      }
```
However we want to apply source presence tracking recursively:
```
if(source.hasFieldB() && source.getFieldB().hasFieldC()) {
        targetObj.setFieldC(fieldC);
      }
```

Is this some feature you can support? Thanks!

**Comment by filiphr** [comment]
How does the `sourceFieldBFieldC` method looks like? 

What you are asking is quite complex to achieve as that can blow up. 

I think that in this case we should not do the `hasFieldB` check, but just check if the result  `field` is not `null`. As the has checks should have already been done in the nested source mapping method.

**Comment by birdfriend** [comment]
auto-generated `sourceFieldBFieldC` looks like:
```
  private String sourceFieldBFieldC(SourceA sourceA) {
    if(sourceA == null) {
      return null;
    } else if(!sourceA.hasFieldB()) {
      return null;
    } else {
      SourceB fieldB = sourceA.getFieldB();
      if(!fieldB.hasFieldC()) {
        return null;
      } else {
        String fieldC = fieldB.getFieldC();
        return fieldC;
      }
    }
  }
```

`sourceFieldBFieldC` has done presence checking correctly. However it returns `null` when the field is not present, and it's possible that the field is present but the value is null.

**Comment by filiphr** [comment]
You are completely right. I would say that if something like #1048 is added to MapStruct then it would work correctly for you.

In theory something like:

```java
if(source.hasFieldB() && source.getFieldB().hasFieldC()) {
     targetObj.setFieldC(fieldC);
}
```

is possible to be added, but it looks really ugly and can blow up if you have more nested levels. On top of that double checks would be done.

**Comment by birdfriend** [comment]
@filiphr thanks for looking into this issue. 
The current behavior on checking presence for nested objects does not seem right though. Before it gets fixed in more elegant way in #1048, can we apply some quick patch to apply source presence tracking for all nested levels?  I agree that it looks ugly, but it works right. 

I can open a pr for the following commit if the change makes sense:
https://github.com/birdfriend/mapstruct/commit/1d0a145ce2c74fb73c761b8663b1c42d6169f269

Thanks!

**Comment by sjaakd** [comment]
closed by merging PR
